client/pyscripts/fm11rf08_full.py

```diff
@@ -0,0 +1,942 @@
+'''
+#==============================================================================
+def readBlocks():
+	global  data
+	global  blkn
+
+	data = []
+	blkn = list(range(0, 63+1)) + list(range(128, 135+1))
+
+	# The user   uses keyhole #1 (-a)
+	# The vendor uses keyhole #2 (-b)
+	# The thief  uses keyhole #4 (backdoor)
+	#                   |___
+	rdbl = f"hf mf rdbl -c 4 --key {dkey} --blk"
+
+	lprint(prompt)
+	lprint(prompt + " Load blocks {0..63, 128..135}[64+8=72] from the card")
+
+	bad = 0
+	for n in blkn:
+		cmd = f"{rdbl} {n}"
+		print(f"\r{prompt} `{cmd}`", end='', flush=True)
+
+		for retry in range(5):
+			p.console(f"{cmd}")
+
+			found = False
+			for line in p.grabbed_output.split('\n'):
+				if " | " in line and "# | s" not in line:
+					l = line[4:76]
+					data.append(l)
+					found = True
+			if found:  break
+
+		if not found:
+			data.append(f"{n:3d} | -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- | ----------------")
+			bad += 1
+
+	print(" .. OK")
+
+#+=============================================================================
+# Patch keys in to data
+# >> "key[][]"
+# >> "data[]"
+# >> keyok!
+'''
+  3 | 00 00 00 00 00 00 87 87 87 69 00 00 00 00 00 00 | .........i......
+'''
+#==============================================================================
+def patchKeys(keyok):
+	global  key
+	global  data
+
+	lprint(prompt)
+	lprint(f"{prompt} Patch keys in to data")
+
+	for sec in range(0, 16+1):
+		blk = (sec * 4) +3  # find "trailer" for this sector
+		if keyok:
+			if key[sec][0] == "":
+				keyA = "-- -- -- -- -- -- "
+			else:
+				kstr = key[sec][0].hex()
+				keyA = "".join([kstr[i:i+2] + " " for i in range(0, len(kstr), 2)])
+
+			if key[sec][1] == "":
+				keyB = "-- -- -- -- -- -- "
+			else:
+				kstr = key[sec][1].hex()
+				keyB = "".join([kstr[i:i+2] + " " for i in range(0, len(kstr), 2)])
+
+			data[blk] = data[blk][:6] + keyA + data[blk][24:36] + keyB
+
+		else:
+			data[blk] = data[blk][:6] + "-- -- -- -- -- -- " + data[blk][24:36] + "-- -- -- -- -- --"
+
+#+=============================================================================
+# Dump data
+# >> blkn
+# >> "data[]"
+#==============================================================================
+def dumpData():
+	global  blkn
+	global  data
+
+	lprint(prompt)
+	lprint(f"{prompt} ===========")
+	lprint(f"{prompt}  Card Data")
+	lprint(f"{prompt} ===========")
+	lprint(f"{prompt}")
+
+	cnt = 0
+	for n in blkn:
+		sec = (cnt // 4)
+		if sec > 15:  sec = sec + 16
+
+		if (n % 4 == 0):
+			lprint(f"{prompt} {sec:2d}:{data[cnt]}")
+		else:
+			lprint(f"{prompt}   :{data[cnt]}")
+
+		cnt += 1
+		if (cnt % 4 == 0) and (n != blkn[-1]):  # Space between sectors
+			lprint(prompt)
+
+#+=============================================================================
+# Let's try to detect a Bambu card by the date strings...
+#==============================================================================
+def detectBambu():
```

Comment:
it would be interesting to isolate this function in a separate file to be imported, so other ppl can use it in their scripts s well, or call the script with a provided block0

Reply:
qv. https://github.com/RfidResearchGroup/proxmark3/pull/2613#discussion_r1830270895